pkg/master/thirdparty/tprregistration_controller.go

```diff
@@ -0,0 +1,214 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kube-aggregator/pkg/apis/apiregistration"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	informers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion/extensions/internalversion"
+	listers "k8s.io/kubernetes/pkg/client/listers/extensions/internalversion"
+	"k8s.io/kubernetes/pkg/registry/extensions/thirdpartyresourcedata"
+)
+
+// AutoAPIServiceRegistration is an interface which callers can re-declare locally and properly cast to for
+// adding and removing APIServices
+type AutoAPIServiceRegistration interface {
+	// AddAPIServiceToSync adds an API service to auto-register.
+	AddAPIServiceToSync(in *apiregistration.APIService)
+	// RemoveAPIServiceToSync removes an API service to auto-register.
+	RemoveAPIServiceToSync(name string)
+}
+
+type tprRegistrationController struct {
+	tprLister listers.ThirdPartyResourceLister
+	tprSynced cache.InformerSynced
+
+	apiServiceRegistration AutoAPIServiceRegistration
+
+	syncHandler func(groupVersion schema.GroupVersion) error
+
+	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
+	// this is actually keyed by a groupVersion
+	queue workqueue.RateLimitingInterface
+}
+
+// NewAutoRegistrationController returns a controller which will register TPR GroupVersions with the auto APIService registration
+// controller so they automatically stay in sync.
+func NewAutoRegistrationController(tprInformer informers.ThirdPartyResourceInformer, apiServiceRegistration AutoAPIServiceRegistration) *tprRegistrationController {
+	c := &tprRegistrationController{
+		tprLister:              tprInformer.Lister(),
+		tprSynced:              tprInformer.Informer().HasSynced,
+		apiServiceRegistration: apiServiceRegistration,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "tpr-autoregister"),
+	}
+	c.syncHandler = c.handleTPR
+
+	tprInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			cast := obj.(*extensions.ThirdPartyResource)
+			c.enqueueTPR(cast)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			cast := obj.(*extensions.ThirdPartyResource)
+			c.enqueueTPR(cast)
+		},
+		DeleteFunc: func(obj interface{}) {
+			cast, ok := obj.(*extensions.ThirdPartyResource)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					glog.V(2).Infof("Couldn't get object from tombstone %#v", obj)
+					return
+				}
+				cast, ok = tombstone.Obj.(*extensions.ThirdPartyResource)
+				if !ok {
+					glog.V(2).Infof("Tombstone contained unexpected object: %#v", obj)
+					return
+				}
+			}
+			c.enqueueTPR(cast)
+		},
+	})
+
+	return c
+}
+
+func (c *tprRegistrationController) Run(threadiness int, stopCh <-chan struct{}) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
```

Comment:
> I do not believe as things stand that HandleCrash() prevents the process from crashing. I believe it just makes sure to log the problem prior to the crash. It's "ReallyCrash" global variable would need to be set to false to prevent the crash.

This defer gives the admin the choice.